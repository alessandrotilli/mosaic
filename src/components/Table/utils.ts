export const COLUMN_CHECKBOX_PATH = "checkbox-selection";
export const COLUMN_ROW_ACTIONS_PATH = "row-actions";

export const TOOLBAR_HEIGHT = 64;
export const TOOLBAR_HEIGHT_MOBILE = 56;

export const PAGINATION_TOOLBAR_HEIGHT = 64;
