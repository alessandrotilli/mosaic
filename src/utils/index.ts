import { ChangeEvent, MouseEvent } from "react";

export interface ISubpart {
  label: string;
  value?: Function;
}

export interface ISubpartSuffix {
  label: string;
  suffix: string;
}

export type ISubpartMap<T extends string = string> = {
  [key in T]: ISubpart;
};

export const DATA_CY_SUFFIX_SEPARATOR = "-";

/**
 * Gets composed dataCy for components with subparts
 * @param dataCy the dataCy prefix
 * @param subpart the subpart to consider
 * @param args optional arguments for subpart evaluation
 */
export const getComposedDataCy = (dataCy: string, subpart: ISubpart, args?: any): string => {
  const { label, value } = subpart;
  return `${dataCy}${DATA_CY_SUFFIX_SEPARATOR}${!value ? slugify(label) : value(args)}`;
};

/**
 * Gets all composed dataCy for all listed subparts
 * @param subpartMap the map of all subparts to consider
 */
export const getAllComposedDataCy = (subpartMap: ISubpartMap): ISubpartSuffix[] =>
  Object.values(subpartMap).map(({ label, value }) => ({
    label,
    suffix: !value ? slugify(label) : value(),
  }));

/**
 * Gets an object property by path, if path is nested is executed recursively
 * @param object the source object
 * @param path the desired path
 */
export const getObjectProperty = (object: any, path: string): any => {
  if (!object) {
    return undefined;
  }

  if (!path.includes(".")) {
    return object[path];
  }

  const pathParts = path.split(".");
  const basePath = pathParts[0];
  if (!object[basePath]) {
    return undefined;
  }

  const baseObject = object[basePath];
  return getObjectProperty(baseObject, path.replace(`${basePath}.`, ""));
};

/**
 * Creates slug or kebab-case string from value
 * @param value the value to slugify
 */
export const slugify = (value: string) =>
  value
    .toLowerCase()
    .replace(/\s+/g, "-")
    .replace(/[^\w-]+/g, "");

/**
 * Suppresses default behaviour and propagation of a given event
 * @param event the event to suppress
 */
export const suppressEvent = (event: ChangeEvent<any> | MouseEvent<any, MouseEvent> | null) => {
  if (!event) {
    return;
  }

  event.preventDefault();
  event.stopPropagation();
};
