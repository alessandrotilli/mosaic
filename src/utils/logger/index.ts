/* eslint-disable no-console */
export type ILogLevel = "error" | "info" | "warn";

const MOSAIC_TAG = "[@melfore/mosaic]";

/**
 * Logs message for given level and component
 * @param level the level of the message to log (e.g. "error")
 * @param component the component subject of the log
 * @param message the message of the log
 */
const logger = (level: ILogLevel, component: string, message: string) => {
  const text = `${MOSAIC_TAG} ${component} - ${message}`;

  switch (level) {
    case "error":
      console.error(text);
      return;
    case "info":
      console.info(text);
      return;
    case "warn":
      console.warn(text);
      return;
  }
};

/**
 * Logs message for error level and component
 * @param component the component subject of the log
 * @param message the message of the log
 */
export const logError = (component: string, message: string) => logger("error", component, message);

/**
 * Logs message for info level and component
 * @param component the component subject of the log
 * @param message the message of the log
 */
export const logInfo = (component: string, message: string) => logger("info", component, message);

/**
 * Logs message for warn level and component
 * @param component the component subject of the log
 * @param message the message of the log
 */
export const logWarn = (component: string, message: string) => logger("warn", component, message);
