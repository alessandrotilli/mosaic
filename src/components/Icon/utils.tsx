import React from "react";
import MUIAccountCircleIcon from "@material-ui/icons/AccountCircle";
import MUIAddIcon from "@material-ui/icons/Add";
import MUIAppsIcon from "@material-ui/icons/Apps";
import MUIArrowUpwardIcon from "@material-ui/icons/ArrowUpward";
import MUIBlockIcon from "@material-ui/icons/Block";
import MUIBrightness4Icon from "@material-ui/icons/Brightness4";
import MUIBusinessIcon from "@material-ui/icons/Business";
import MUICheckIcon from "@material-ui/icons/Check";
import MUICheckBoxIcon from "@material-ui/icons/CheckBox";
import MUICheckBoxOutlineBlankIcon from "@material-ui/icons/CheckBoxOutlineBlank";
import MUIChevronLeftIcon from "@material-ui/icons/ChevronLeft";
import MUIChevronRightIcon from "@material-ui/icons/ChevronRight";
import MUICloseIcon from "@material-ui/icons/Close";
import MUICropIcon from "@material-ui/icons/Crop";
import MUIDeleteIcon from "@material-ui/icons/Delete";
import MUIDetailsIcon from "@material-ui/icons/Details";
import MUIDraftsIcon from "@material-ui/icons/Drafts";
import MUIEditIcon from "@material-ui/icons/Edit";
import MUIErrorIcon from "@material-ui/icons/Error";
import MUIFilterListIcon from "@material-ui/icons/FilterList";
import MUIFirstPageIcon from "@material-ui/icons/FirstPage";
import MUIGroupWorkIcon from "@material-ui/icons/GroupWork";
import MUIHomeIcon from "@material-ui/icons/Home";
import MUIKeyboardArrowDownIcon from "@material-ui/icons/KeyboardArrowDown";
import MUIKeyboardArrowUpIcon from "@material-ui/icons/KeyboardArrowUp";
import MUILanguageIcon from "@material-ui/icons/Language";
import MUILastPageIcon from "@material-ui/icons/LastPage";
import MUIListIcon from "@material-ui/icons/List";
import MUIMailIcon from "@material-ui/icons/Mail";
import MUIMenuIcon from "@material-ui/icons/Menu";
import MUINotificationsIcon from "@material-ui/icons/Notifications";
import MUIOpenInNew from "@material-ui/icons/OpenInNew";
import MUIPauseIcon from "@material-ui/icons/Pause";
import MUIPlayArrowIcon from "@material-ui/icons/PlayArrow";
import MUIPlaylistAddCheckIcon from "@material-ui/icons/PlaylistAddCheck";
import MUIPowerSettingsNewIcon from "@material-ui/icons/PowerSettingsNew";
import MUIPrintIcon from "@material-ui/icons/Print";
import MUIRefreshIcon from "@material-ui/icons/Refresh";
import MUISaveIcon from "@material-ui/icons/Save";
import MUIScheduleIcon from "@material-ui/icons/Schedule";
import MUISearchIcon from "@material-ui/icons/Search";
import MUISendIcon from "@material-ui/icons/Send";
import MUISettingsIcon from "@material-ui/icons/Settings";
import MUISignalWifi1BarOutlinedIcon from "@material-ui/icons/SignalWifi1BarOutlined";
import MUITimelineIcon from "@material-ui/icons/Timeline";
import MUIVisibilityIcon from "@material-ui/icons/Visibility";
import MUIVisibilityOffIcon from "@material-ui/icons/VisibilityOff";
import MUIWarningIcon from "@material-ui/icons/Warning";
import MUIWifiOffOutlinedIcon from "@material-ui/icons/WifiOffOutlined";
import MUIWifiOutlinedIcon from "@material-ui/icons/WifiOutlined";

import { Icons, IIconsCatalog } from "../../types/Icon";

export const iconsCatalog: IIconsCatalog = {
  [Icons.account]: <MUIAccountCircleIcon />,
  [Icons.add]: <MUIAddIcon />,
  [Icons.apps]: <MUIAppsIcon />,
  [Icons.block]: <MUIBlockIcon />,
  [Icons.brightness]: <MUIBrightness4Icon />,
  [Icons.business]: <MUIBusinessIcon />,
  [Icons.chart_timeline]: <MUITimelineIcon />,
  [Icons.check]: <MUICheckIcon />,
  [Icons.checkbox]: <MUICheckBoxIcon />,
  [Icons.checkbox_empty]: <MUICheckBoxOutlineBlankIcon />,
  [Icons.clock]: <MUIScheduleIcon />,
  [Icons.close]: <MUICloseIcon />,
  [Icons.connection_slow]: <MUISignalWifi1BarOutlinedIcon />,
  [Icons.crop]: <MUICropIcon />,
  [Icons.delete]: <MUIDeleteIcon />,
  [Icons.details]: <MUIDetailsIcon />,
  [Icons.down]: <MUIKeyboardArrowDownIcon />,
  [Icons.edit]: <MUIEditIcon />,
  [Icons.error]: <MUIErrorIcon />,
  [Icons.filter]: <MUIFilterListIcon />,
  [Icons.first]: <MUIFirstPageIcon />,
  [Icons.group]: <MUIGroupWorkIcon />,
  [Icons.home]: <MUIHomeIcon />,
  [Icons.language]: <MUILanguageIcon />,
  [Icons.last]: <MUILastPageIcon />,
  [Icons.left]: <MUIChevronLeftIcon />,
  [Icons.list_check]: <MUIPlaylistAddCheckIcon />,
  [Icons.list]: <MUIListIcon />,
  [Icons.mail_open]: <MUIDraftsIcon />,
  [Icons.mail]: <MUIMailIcon />,
  [Icons.menu]: <MUIMenuIcon />,
  [Icons.notifications]: <MUINotificationsIcon />,
  [Icons.open_new]: <MUIOpenInNew />,
  [Icons.pause]: <MUIPauseIcon />,
  [Icons.play]: <MUIPlayArrowIcon />,
  [Icons.power]: <MUIPowerSettingsNewIcon />,
  [Icons.print]: <MUIPrintIcon />,
  [Icons.refresh]: <MUIRefreshIcon />,
  [Icons.right]: <MUIChevronRightIcon />,
  [Icons.save]: <MUISaveIcon />,
  [Icons.search]: <MUISearchIcon />,
  [Icons.send]: <MUISendIcon />,
  [Icons.settings]: <MUISettingsIcon />,
  [Icons.up_arrow]: <MUIArrowUpwardIcon />,
  [Icons.up]: <MUIKeyboardArrowUpIcon />,
  [Icons.visibility]: <MUIVisibilityIcon />,
  [Icons.visibility_off]: <MUIVisibilityOffIcon />,
  [Icons.warn]: <MUIWarningIcon />,
  [Icons.wifi_off]: <MUIWifiOffOutlinedIcon />,
  [Icons.wifi_on]: <MUIWifiOutlinedIcon />,
};
