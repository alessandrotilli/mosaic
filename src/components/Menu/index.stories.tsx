import React from "react";
import MUIStyleIcon from "@mui/icons-material/Style";
import { expect, jest } from "@storybook/jest";
import { Meta, StoryObj } from "@storybook/react";
import { configure, userEvent, within } from "@storybook/testing-library";

import { logInfo } from "../../utils/logger";
import { localeDecorator } from "../../utils/mocks/LocaleMock";
import getDocsPage from "../../utils/stories";

import Menu, { DATA_CY_DEFAULT } from ".";

configure({ testIdAttribute: "data-cy" });

const COMPONENT_NAME = Menu.displayName!;

const meta = {
  title: "Navigation/Menu",
  component: Menu,
  decorators: [localeDecorator],
  parameters: {
    docs: {
      ...getDocsPage({
        basedOn: {
          label: "MUI Menu Component",
          url: "https://mui.com/material-ui/react-menu/",
        },
        component: COMPONENT_NAME,
        e2eTestInfo: {
          dataCyDefault: DATA_CY_DEFAULT,
        },
      }),
    },
  },
} satisfies Meta<typeof Menu>;

export default meta;
type Story = StoryObj<typeof meta>;

const onClickMock = jest.fn(() => logInfo(COMPONENT_NAME, "onClick handler"));

const onItemClickMock = jest.fn(() => logInfo(COMPONENT_NAME, "onClick item handler"));

export const Primary: Story = {
  args: {
    items: [{ label: "Item", onClick: onItemClickMock, value: "Value" }],
    label: "Label",
    onItemClick: onClickMock,
  },
  play: async ({ canvasElement, step }) => {
    const canvas = within(canvasElement);
    const button = canvas.getByTestId(`${DATA_CY_DEFAULT}-button`);
    if (!button) {
      return;
    }

    await step("Open Menu", async () => {
      await userEvent.click(button);
      await expect(onClickMock).toHaveBeenCalledTimes(onClickMock.mock.calls.length);
    });

    // await step("Click Menu Item", async () => {
    //   const item = canvas.getByTestId(`${DATA_CY_DEFAULT}-item-0`);
    //   if (!item) {
    //     return;
    //   }

    //   await userEvent.click(item);
    //   await expect(onItemClickMock).toHaveBeenCalledTimes(onItemClickMock.mock.calls.length);
    // });
  },
};

export const CustomIcon: Story = {
  args: {
    ...Primary.args,
    icon: <MUIStyleIcon />,
  },
};

export const IconButton: Story = {
  args: {
    dataCy: "icon-menu",
    items: [{ label: "Item", value: "Value" }],
    label: "Label",
    onItemClick: onClickMock,
    type: "icon",
  },
};

export const CustomIconButton: Story = {
  args: {
    ...IconButton.args,
    icon: <MUIStyleIcon />,
  },
};
