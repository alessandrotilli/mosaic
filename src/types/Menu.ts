import { IBase } from "./Base";
import { IPartialIconUtilizer } from "./Icon";

export type IMenuItemCallback = (value: string) => void;

export interface IMenuButton extends IBase, IPartialIconUtilizer {
  /**
   * Menu label
   */
  label: string;
  /**
   * Menu button type 'button' or 'icon'
   */
  type?: IMenuType;
}

export interface IMenuItem extends IBase {
  /**
   * Menu item label
   */
  label: string;
  /**
   * Callback for click events, dedicated for menu item
   */
  onClick?: IMenuItemCallback;
  /**
   * Value of menu item
   */
  value: string;
}

export type IMenuType = "button" | "icon";

export interface IMenu extends IBase, IMenuButton {
  /**
   * List of menu items
   */
  items: IMenuItem[];
  /**
   * Callback for click events, applied to all menu items
   */
  onItemClick?: IMenuItemCallback;
}
