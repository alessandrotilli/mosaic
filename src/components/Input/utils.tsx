import React from "react";
import { InputAdornment as MUIInputAdornment } from "@material-ui/core";

import { IInputAdornment } from "../../types/Input";
import IconButton from "../IconButton";

export const getAdornment = (adornment?: IInputAdornment) => {
  if (!adornment) {
    return;
  }

  const { icon, onClick } = adornment;
  return (
    <MUIInputAdornment position="end">
      <IconButton disabled={!onClick} icon={icon} onClick={onClick!} size="small" />
    </MUIInputAdornment>
  );
};
