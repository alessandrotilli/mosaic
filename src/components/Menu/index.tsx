import React, { FC, Fragment, useCallback, useMemo, useState } from "react";
import { Menu as MUIMenu, MenuItem as MUIMenuItem, PopoverOrigin as MUIPopoverOrigin } from "@material-ui/core";

import { Icons } from "../../types/Icon";
import { IMenu, IMenuItemCallback } from "../../types/Menu";
import { suppressEvent } from "../../utils";
import { logWarn } from "../../utils/logger";
import Button from "../Button";
import IconButton from "../IconButton";

const MENU_ITEMS_ANCHORING: MUIPopoverOrigin = {
  vertical: "top",
  horizontal: "right",
};

const Menu: FC<IMenu> = ({ dataCy, icon: externalIcon, label, items, onItemClick, style }) => {
  const [anchor, setAnchor] = useState<Element | null>(null);

  const onClose = useCallback(() => setAnchor(null), []);

  const onItem = useCallback(
    (value: string, onClick?: IMenuItemCallback) => {
      if (onClick) {
        onClick(value);
        return;
      }

      if (onItemClick) {
        onItemClick(value);
        return;
      }

      return;
    },
    [onItemClick]
  );

  const onMenu = useCallback(() => setAnchor(document.querySelector(`button[data-cy='${dataCy}']`)), [dataCy]);

  const menuButton = useMemo(() => {
    let icon = externalIcon;
    if (!icon) {
      logWarn("Menu", "Icon is not set, fallback to Icons.menu");
      icon = Icons.menu;
    }

    if (!label) {
      return <IconButton dataCy={dataCy} icon={icon} onClick={onMenu} style={style} />;
    }

    return (
      <Button
        dataCy={dataCy}
        icon={typeof icon === "string" ? { name: icon } : { component: icon }}
        label={label}
        onClick={onMenu}
        style={style}
      />
    );
  }, [dataCy, externalIcon, label, onMenu, style]);

  return (
    <Fragment>
      {menuButton}
      <MUIMenu
        id={dataCy}
        anchorEl={anchor}
        anchorOrigin={MENU_ITEMS_ANCHORING}
        keepMounted
        transformOrigin={MENU_ITEMS_ANCHORING}
        open={!!anchor}
        onClose={onClose}
      >
        {items.map(({ label, onClick, value }, index) => (
          <MUIMenuItem
            key={`${dataCy}-item-${index}`}
            data-cy={`${dataCy}-item-${index}`}
            onClick={(event) => {
              suppressEvent(event);
              onClose();
              onItem(value, onClick);
            }}
          >
            {label}
          </MUIMenuItem>
        ))}
      </MUIMenu>
    </Fragment>
  );
};

export default Menu;
