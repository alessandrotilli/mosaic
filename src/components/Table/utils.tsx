import React, { forwardRef, Fragment } from "react";
import { Action as MTActionType, Column as MTColumnType, Options as MTOptionsType } from "material-table";
import { Icons, IconSize } from "../../types/Icon";
import { TableActionPosition, ITableAction, ITableColumn } from "../../types/Table";
import Button from "../Button";
import Icon from "../Icon";
import IconButton from "../IconButton";
import { getDataCyForSubComponent, suppressEvent } from "../../utils";

// For default props refer to Options section here https://material-table.com/#/docs/all-props
export const DEFAULT_TABLE_OPTIONS: MTOptionsType = {
  debounceInterval: 500,
  draggable: false,
  emptyRowsWhenPaging: false,
};

export const actionAdapter = (action: ITableAction): MTActionType<object> => {
  const {
    callback,
    label,
    disabled = false,
    hidden = false,
    icon = undefined,
    position = TableActionPosition.default,
  } = action;
  return {
    disabled,
    hidden,
    icon: `${label}$${icon}`,
    onClick: (event, data) => {
      if (event) {
        suppressEvent(event);
      }
      callback(data);
    },
    position,
  };
};

export const actionComponentAdapter = (props: any, dataCy: string) => {
  const { data, action } = props;
  const { disabled, hidden, icon, position, onClick }: MTActionType<object> = action;
  if (hidden) {
    return null;
  }

  const displaysButton = position === TableActionPosition.default;
  const labelAndIconName = icon as string;
  const label = labelAndIconName.split("$")[0];
  const iconName = labelAndIconName.split("$")[1] as Icons;
  return (
    <Fragment>
      {displaysButton && (
        <Fragment>
          <span>&nbsp;&nbsp;&nbsp;&nbsp;</span>
          <Button
            dataCy={getDataCyForSubComponent(dataCy, `action-${label}`)}
            disabled={disabled}
            icon={!iconName ? undefined : { name: iconName }}
            label={label}
            onClick={() => onClick(undefined, data)}
          />
        </Fragment>
      )}
      {!displaysButton && (
        <Fragment>
          <IconButton
            dataCy={getDataCyForSubComponent(dataCy, `action-${label}`)}
            disabled={disabled}
            icon={iconName}
            onClick={() => onClick(undefined, data)}
            size={IconSize.small}
          />
          <span>&nbsp;&nbsp;</span>
        </Fragment>
      )}
    </Fragment>
  );
};

export const columnAdapter = (column: ITableColumn): MTColumnType<object> => {
  const { label, path, render, width } = column;
  return {
    ...(!!width && ({ width } as object)),
    field: path,
    render,
    title: label,
  };
};

export const iconAdapter = (name: Icons, size = IconSize.default) =>
  forwardRef((props: any, ref: any) => {
    return <Icon forwarded={{ ...props, ref }} name={name} size={size} />;
  });
