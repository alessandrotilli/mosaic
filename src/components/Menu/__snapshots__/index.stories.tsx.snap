// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navigation/Menu Primary smoke-test 1`] = `
<button class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-disableElevation MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-disableElevation css-zddlty-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="button"
        data-cy="Menu"
>
  <span class="MuiButton-startIcon MuiButton-iconSizeMedium css-1d6wzja-MuiButton-startIcon">
    <svg class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1wnp694-MuiSvgIcon-root"
         focusable="false"
         aria-hidden="true"
         viewbox="0 0 24 24"
         data-testid="MenuIcon"
         data-cy="Menu-icon"
    >
      <path d="M3 18h18v-2H3v2zm0-5h18v-2H3v2zm0-7v2h18V6H3z">
      </path>
    </svg>
  </span>
  Label
  <span class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root">
  </span>
</button>
`;
